File: odk/examples/DevelopersGuide/FirstSteps/HelloTextTableShape/python/HelloTextTableShape.py
# -*- tab-width: 4; indent-tabs-mode: nil; py-indent-offset: 4 -*-
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#

import sys
import traceback

import officehelper

from com.sun.star.lang import DisposedException
from com.sun.star.awt.FontSlant import ITALIC
from com.sun.star.awt.FontWeight import BOLD
from com.sun.star.table import BorderLine
from com.sun.star.table import TableBorder
from com.sun.star.awt import Size
from com.sun.star.awt import Point


class HelloTextTableShape:

    def __init__(self):
        self.remote_context = None
        self.remote_service_manager = None

    def use_documents(self) -> None:
        self.use_writer()
        self.use_calc()
        self.use_draw()

    def get_remote_service_manager(self) -> None:
        try:
            self.remote_context = officehelper.bootstrap()
            print("Connected to a running office ...")
            return self.remote_context.ServiceManager
        except Exception:
            traceback.print_exc()
            sys.exit(1)

    def new_doc_component(self, doc_type: str) -> None:
        load_url = "private:factory/" + doc_type
        self.remote_service_manager = self.get_remote_service_manager()
        desktop = self.remote_service_manager.createInstanceWithContext(
            "com.sun.star.frame.Desktop", self.remote_context
        )
        return desktop.loadComponentFromURL(load_url, "_blank", 0, tuple([]))

    def use_writer(self) -> None:
        try:
            doc = self.new_doc_component("swriter")
            xtext = doc.Text
            self.manipulateText(xtext)

            # insert TextTable and get cell text, then manipulate text in cell
            table = doc.createInstance("com.sun.star.text.TextTable")
            xtext.insertTextContent(xtext.End, table, False)

            xcell = table[1, 0]
            self.manipulateText(xcell)
            self.manipulateTable(table)

            # insert RectangleShape and get shape text, then manipulate text
            writer_shape = doc.createInstance("com.sun.star.drawing.RectangleShape")
            writer_shape.setSize(Size(10000, 10000))
            xtext.insertTextContent(xtext.End, writer_shape, False)
            # wrap text inside shape
            writer_shape.TextContourFrame = True

            self.manipulateText(writer_shape)
            self.manipulateShape(writer_shape)

            bookmark = doc.createInstance("com.sun.star.text.Bookmark")
            bookmark.Name = "MyUniqueBookmarkName"
            # insert the bookmark at the end of the document
            xtext.insertTextContent(xtext.End, bookmark, False)

            # Query the added bookmark and set a string
            found_bookmark = doc.Bookmarks.getByName("MyUniqueBookmarkName")
            found_bookmark.Anchor.String = (
                "The throat mike, glued to her neck, "
                "looked as much as possible like an analgesic dermadisk."
            )

            for text_table in doc.TextTables:
                text_table.BackColor = 0xC8FFB9
        except DisposedException:
            self.remote_context = None
            raise

    def use_calc(self) -> None:
        try:
            doc = self.new_doc_component("scalc")
            sheet = doc.Sheets[0]

            # get cell A2 in first sheet
            cell = sheet[1, 0]
            cell.IsTextWrapped = True

            self.manipulateText(cell)
            self.manipulateTable(sheet)

            # create and insert RectangleShape and get shape text,
            # then manipulate text
            shape = doc.createInstance("com.sun.star.drawing.RectangleShape")
            shape.Size = Size(10000, 10000)
            shape.Position = Point(7000, 3000)
            shape.TextContourFrame = True
            sheet.DrawPage.add(shape)

            self.manipulateText(shape)
            self.manipulateShape(shape)
        except DisposedException:
            self.remote_context = None
            raise

    def use_draw(self) -> None:
        try:
            doc = self.new_doc_component("sdraw")

            draw_shape = doc.createInstance("com.sun.star.drawing.RectangleShape")
            draw_shape.setSize(Size(10000, 20000))
            draw_shape.setPosition(Point(5000, 5000))
            doc.DrawPages[0].add(draw_shape)

            # wrap text inside shape
            draw_shape.TextContourFrame = True

            self.manipulateText(draw_shape)
            self.manipulateShape(draw_shape)
        except DisposedException:
            self.remote_context = None
            raise

    def manipulateText(self, xtext) -> None:
        """Insert text content

        :param xtext: object that implements com.sun.star.text.XText interface.
        """
        # simply set whole text as one string
        xtext.String = (
            "He lay flat on the brown, pine-needled floor of the forest, "
            "his chin on his folded arms, and high overhead the wind blew "
            "in the tops of the pine trees."
        )

        # create text cursor for selecting and formatting
        text_cursor = xtext.createTextCursor()
        # use cursor to select "He lay" and apply bold italic
        text_cursor.gotoStart(False)
        text_cursor.goRight(6, True)
        # from CharacterProperties
        text_cursor.CharPosture = ITALIC
        text_cursor.CharWeight = BOLD

        # add more text at the end of the text using insertString
        text_cursor.gotoEnd(False)
        content = (
            " The mountainside sloped gently where he lay; "
            "but below it was steep and he could see the dark of the oiled "
            "road winding through the pass. There was a stream alongside the "
            "road and far down the pass he saw a mill beside the stream and "
            "the falling water of the dam, white in the summer sunlight."
        )
        xtext.insertString(text_cursor, content, False)
        # after insertString the cursor is behind the inserted text,
        # insert more text
        content = "\n  \"Is that the mill?\" he asked."
        xtext.insertString(text_cursor, content, False)

    def manipulateTable(self, xcellrange) -> None:
        """Format a table area

        :param xcellrange: object that implements com.sun.star.table.XCellRange interface.
        """
        # enter column titles and a cell value
        xcellrange[0, 0].String = "Quotation"
        xcellrange[0, 1].String = "Year"
        xcellrange[1, 1].Value = 1940

        # format table headers and table borders
        # we need to distinguish text and sheet tables:
        # property name for cell colors is different in text and sheet cells
        # we want to apply TableBorder to whole text table, but only to sheet
        # cells with content

        background_color = 0x99CCFF

        # create description for blue line, width 10
        line = BorderLine()
        line.Color = 0x000099
        line.OuterLineWidth = 10
        # apply line description to all border lines and make them valid
        border = TableBorder()
        border.VerticalLine = border.HorizontalLine = line
        border.LeftLine = border.RightLine = line
        border.TopLine = border.BottomLine = line
        border.IsVerticalLineValid = border.IsHorizontalLineValid = True
        border.IsLeftLineValid = border.IsRightLineValid = True
        border.IsTopLineValid = border.IsBottomLineValid = True

        if xcellrange.supportsService("com.sun.star.sheet.Spreadsheet"):
            selected_cells = xcellrange["A1:B2"]
            selected_cells.CellBackColor = background_color
            selected_cells.TableBorder = border
            print(selected_cells.TableBorder.TopLine.Color)
        elif xcellrange.supportsService("com.sun.star.text.TextTable"):
            selected_cells = xcellrange["A1:B1"]
            selected_cells.BackColor = background_color
            xcellrange.TableBorder = border
            print(xcellrange.TableBorder.TopLine.Color)

    def manipulateShape(self, xshape) -> None:
        """Format a shape

        :param xshape: object that implements com.sun.star.drawing.XShape interface.
        """
        xshape.FillColor = 0x99CCFF
        xshape.LineColor = 0x000099
        xshape.RotateAngle = 3000

        xshape.TextLeftDistance = 0
        xshape.TextRightDistance = 0
        xshape.TextUpperDistance = 0
        xshape.TextLowerDistance = 0


def main() -> None:
    try:
        HelloTextTableShape().use_documents()
    except Exception as e:
        print(str(e))
        traceback.print_exc()


if __name__ == "__main__":
    main()

# vim: set shiftwidth=4 softtabstop=4 expandtab:

File: odk/examples/DevelopersGuide/FirstSteps/HelloTextTableShape/csharp/HelloTextTableShape.cs
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

using System;

using com.sun.star.awt;
using com.sun.star.beans;
using com.sun.star.container;
using com.sun.star.drawing;
using com.sun.star.frame;
using com.sun.star.lang;
using com.sun.star.sheet;
using com.sun.star.table;
using com.sun.star.text;
using com.sun.star.uno;

XComponentContext xRemoteContext = null;
XMultiComponentFactory xRemoteServiceManager = null;

try
{
    UseWriter();
    UseCalc();
    UseDraw();
    return 0;
}
catch (UnoException e)
{
    Console.Error.WriteLine(e.Message);
    return 1;
}

void UseWriter()
{
    // Create a new Writer document, and manipulate the text content
    XComponent xWriterComponent = NewDocComponent("swriter");
    XTextDocument xTextDocument = xWriterComponent.query<XTextDocument>();
    XText xText = xTextDocument.getText();

    ManipulateText(xText);

    // Get the internal service factory of the document
    XMultiServiceFactory xWriterFactory = xTextDocument.query<XMultiServiceFactory>();

    // Insert a TextTable, and manipulate the text content of the cell
    XTextContent xTextContentTable = xWriterFactory.createInstance("com.sun.star.text.TextTable").query<XTextContent>();
    xText.insertTextContent(xText.getEnd(), xTextContentTable, false);

    XCellRange xCellRange = xTextContentTable.query<XCellRange>();
    XCell xCell = xCellRange.getCellByPosition(0, 1);
    XText xCellText = xCell.query<XText>();

    ManipulateText(xCellText);
    ManipulateTable(xCellRange);

    XShape xWriterShape = xWriterFactory.createInstance("com.sun.star.drawing.RectangleShape").query<XShape>();
    xWriterShape.setSize(new Size(10000, 10000));
    XTextContent xTextContentShape = xWriterShape.query<XTextContent>();

    xText.insertTextContent(xText.getEnd(), xTextContentShape, false);

    // Wrap the text inside the shape
    XPropertySet xShapeProps = xWriterShape.query<XPropertySet>();
    xShapeProps.setPropertyValue("TextContourFrame", new Any(true));

    XText xShapeText = xWriterShape.query<XText>();

    ManipulateText(xShapeText);
    ManipulateShape(xWriterShape);

    // Create and name a bookmark
    IQueryInterface bookmark = xWriterFactory.createInstance("com.sun.star.text.Bookmark");
    XNamed xNamed = bookmark.query<XNamed>();
    xNamed.setName("MyUniqueBookmarkName");

    // Insert the bookmark at the end of the document
    XTextContent xTextContent = bookmark.query<XTextContent>();
    xText.insertTextContent(xText.getEnd(), xTextContent, false);

    // Get all bookmarks from the XBookmarksSupplier
    XBookmarksSupplier xBookmarksSupplier = xWriterComponent.query<XBookmarksSupplier>();
    XNameAccess xNamedBookmarks = xBookmarksSupplier.getBookmarks();
    XTextContent xFoundBookmark = xNamedBookmarks.getByName("MyUniqueBookmarkName").cast<XTextContent>();
    XTextRange xFound = xFoundBookmark.getAnchor();
    xFound.setString(" The throat mike, glued to her neck, "
        + "looked as much as possible like an analgesic dermadisk.");

    // Get all tables from the XTextTablesSupplier
    XTextTablesSupplier xTablesSupplier = xWriterComponent.query<XTextTablesSupplier>();
    XNameAccess xNamedTables = xTablesSupplier.getTextTables();
    XIndexAccess xIndexedTables = xNamedTables.query<XIndexAccess>();

    // Get the tables
    XPropertySet xTableProps = null;
    for (int i = 0; i < xIndexedTables.getCount(); i++)
    {
        xTableProps = xIndexedTables.getByIndex(i).cast<XPropertySet>();
        xTableProps.setPropertyValue("BackColor", new Any(0xC8FFB9));
    }
}

void UseCalc()
{
    // Create a new Calc document, and manipulate text in a cell
    XComponent xCalcComponent = NewDocComponent("scalc");
    XSpreadsheetDocument xCalcDocument = xCalcComponent.query<XSpreadsheetDocument>();
    XSpreadsheetDocument xSpreadsheetDocument = xCalcDocument.query<XSpreadsheetDocument>();

    XIndexAccess xIndexedSheets = xSpreadsheetDocument.getSheets().query<XIndexAccess>();
    XCellRange xSpreadsheetCells = xIndexedSheets.getByIndex(0).cast<XCellRange>();

    // Get cell A2 from the first sheet
    XCell xCell = xSpreadsheetCells.getCellByPosition(0, 1);
    XText xCellText = xCell.query<XText>();

    XPropertySet xCellProps = xCell.query<XPropertySet>();
    xCellProps.setPropertyValue("IsTextWrapped", new Any(true));

    ManipulateText(xCellText);
    ManipulateTable(xSpreadsheetCells);

    // Get the internal service factory of the document
    XMultiServiceFactory xCalcFactory = xCalcDocument.query<XMultiServiceFactory>();
    XDrawPageSupplier xDrawPageSupplier = xIndexedSheets.getByIndex(0).cast<XDrawPageSupplier>();
    XDrawPage xDrawPage = xDrawPageSupplier.getDrawPage();

    // Create and insert a RectangleShape, and manipulate its shape text
    XShape xCalcShape = xCalcFactory.createInstance("com.sun.star.drawing.RectangleShape").query<XShape>();
    xCalcShape.setSize(new Size(10000, 10000));
    xCalcShape.setPosition(new Point(7000, 3000));
    xDrawPage.add(xCalcShape);

    // Wrap the text inside the shape
    XPropertySet xShapeProps = xCalcShape.query<XPropertySet>();
    xShapeProps.setPropertyValue("TextContourFrame", new Any(true));

    XText xShapeText = xCalcShape.query<XText>();
    ManipulateText(xShapeText);
    ManipulateShape(xCalcShape);
}

void UseDraw()
{
    // Create a new Draw document, and insert a rectangle
    XComponent xDrawComponent = NewDocComponent("sdraw");
    XDrawPagesSupplier xDrawPagesSupplier = xDrawComponent.query<XDrawPagesSupplier>();

    XIndexAccess xIndexedDrawPages = xDrawPagesSupplier.getDrawPages();
    XDrawPage xDrawPage = xIndexedDrawPages.getByIndex(0).cast<XDrawPage>();

    // Get the internal service factory of the document
    XMultiServiceFactory xDrawFactory = xDrawComponent.query<XMultiServiceFactory>();
    XShape xDrawShape = xDrawFactory.createInstance("com.sun.star.drawing.RectangleShape").query<XShape>();
    xDrawShape.setSize(new Size(10000, 20000));
    xDrawShape.setPosition(new Point(5000, 5000));
    xDrawPage.add(xDrawShape);

    // Wrap the text inside the shape
    XPropertySet xShapeProps = xDrawShape.query<XPropertySet>();
    xShapeProps.setPropertyValue("TextContourFrame", new Any(true));

    XText xShapeText = xDrawShape.query<XText>();
    ManipulateText(xShapeText);
    ManipulateShape(xDrawShape);
}

void ManipulateText(XText xText)
{
    // Set the entire text at once
    xText.setString("He lay flat on the brown, pine-needled floor of the forest, "
        + "his chin on his folded arms, and high overhead the wind blew in the tops "
        + "of the pine trees.");

    // Create a text cursor for selecting and formatting
    XTextCursor xTextCursor = xText.createTextCursor();
    XPropertySet xCursorProps = xTextCursor.query<XPropertySet>();

    // Use the cursor to select "He lay", and apply bold and italic formatting
    xTextCursor.gotoStart(false);
    xTextCursor.goRight(6, true);

    xCursorProps.setPropertyValue("CharPosture", new Any(FontSlant.ITALIC));
    xCursorProps.setPropertyValue("CharWeight", new Any(FontWeight.BOLD));

    xTextCursor.gotoEnd(false);
    xText.insertString(xTextCursor.getStart(),
        " The mountainside sloped gently where he lay; "
        + "but below it was steep and he could see the dark of the oiled road "
        + "winding through the pass. There was a stream alongside the road "
        + "and far down the pass he saw a mill beside the stream and the falling water "
        + "of the dam, white in the summer sunlight.", false);
    xText.insertString(xTextCursor.getStart(), "\n  \"Is that the mill?\" he asked.", false);
}

void ManipulateTable(XCellRange xCellRange)
{
    // Set column titles and a cell value
    XCell xCell = xCellRange.getCellByPosition(0, 0);
    XText xCellText = xCell.query<XText>();
    xCellText.setString("Quotation");

    xCell = xCellRange.getCellByPosition(1, 0);
    xCellText = xCell.query<XText>();
    xCellText.setString("Year");

    xCell = xCellRange.getCellByPosition(1, 1);
    xCell.setValue(1940);

    XCellRange xSelectedCells = xCellRange.getCellRangeByName("A1:B1");
    XPropertySet xCellProps = xSelectedCells.query<XPropertySet>();

    string backColorPropertyName = null;
    XPropertySet xTableProps = null;

    // Format the table headers and borders
    XServiceInfo xServiceInfo = xCellRange.query<XServiceInfo>();
    if (xServiceInfo.supportsService("com.sun.star.sheet.Spreadsheet"))
    {
        backColorPropertyName = "CellBackColor";
        xSelectedCells = xCellRange.getCellRangeByName("A1:B2");
        xTableProps = xSelectedCells.query<XPropertySet>();
    }
    else if (xServiceInfo.supportsService("com.sun.star.text.TextTable"))
    {
        backColorPropertyName = "BackColor";
        xTableProps = xCellRange.query<XPropertySet>();
    }
    // Set cell background color
    xCellProps.setPropertyValue(backColorPropertyName, new Any(0x99CCFF));

    // Set table borders (blue line, width 10)
    BorderLine theLine = new BorderLine()
    {
        Color = 0x000099,
        OuterLineWidth = 10,
    };
    
    // Use the above defined line for all borders and enable them
    TableBorder bord = new TableBorder()
    {
        VerticalLine = theLine,
        HorizontalLine = theLine,
        LeftLine = theLine,
        RightLine = theLine,
        TopLine = theLine,
        BottomLine = theLine,

        IsVerticalLineValid = true,
        IsHorizontalLineValid = true,
        IsLeftLineValid = true,
        IsRightLineValid = true,
        IsTopLineValid = true,
        IsBottomLineValid = true,
    };
    xTableProps.setPropertyValue("TableBorder", new Any(bord));

    bord = xTableProps.getPropertyValue("TableBorder").cast<TableBorder>();
    Console.WriteLine(bord.TopLine.Color);
}

void ManipulateShape(XShape xShape)
{
    XPropertySet xShapeProps = xShape.query<XPropertySet>();
    xShapeProps.setPropertyValue("FillColor", new Any(0x99CCFF));
    xShapeProps.setPropertyValue("LineColor", new Any(0x000099));
    xShapeProps.setPropertyValue("RotateAngle", new Any(3000));
    xShapeProps.setPropertyValue("TextLeftDistance", new Any(0));
    xShapeProps.setPropertyValue("TextRightDistance", new Any(0));
    xShapeProps.setPropertyValue("TextUpperDistance", new Any(0));
    xShapeProps.setPropertyValue("TextLowerDistance", new Any(0));
}

XComponent NewDocComponent(string docType)
{
    XMultiComponentFactory xRemoteServiceManager = GetRemoteServiceManager();

    XComponentLoader xComponentLoader = xRemoteServiceManager
        .createInstanceWithContext("com.sun.star.frame.Desktop", xRemoteContext)
        .query<XComponentLoader>();

    string loadUrl = $"private:factory/{docType}";
    return xComponentLoader.loadComponentFromURL(loadUrl, "_blank", 0, new PropertyValue[0]);
}

XMultiComponentFactory GetRemoteServiceManager()
{
    if (xRemoteContext == null && xRemoteServiceManager == null)
    {
        xRemoteContext = NativeBootstrap.bootstrap();
        Console.WriteLine("Connected to a running office ...");

        xRemoteServiceManager = xRemoteContext.getServiceManager();
    }
    return xRemoteServiceManager;
}